### PR TITLE
Do not consider InUse registers for assigning to consecutive candidates

### PR DESCRIPTION
**Pull request by kunalspathak**
When checking if consecutive registers are available, also take into account the information if registers are in use at the same location. If that is the case, do not consider it as candidate for consecutive register allocation.

I also experimented with 6e8b09624165bca37b33336d0c0181e268aa7170 to not spill a register that is already assigned to the consecutive registers because that could involve lot of moving around, since the next consecutive registers might have to be in different registers than the ones they were originally obtained. It doesn't make much difference in the codegen of the example of the code that was introduced in #95513 . I might revert depending on how CI and TP impact is.

image

 
Details: https://github.com/dotnet/runtime/pull/95513#discussion_r1414572949

### CODE REVIEW

**Comment by kunalspathak** [comment]
@dotnet/jit-contrib

File: src/coreclr/jit/lsra.cpp [hunk 1]
```diff
@@ -11949,18 +11949,30 @@ void LinearScan::RegisterSelection::try_SPILL_COST()
...
                     currentSpillWeight = linearScan->getWeight(reloadRefPosition);
                 }
+#ifdef TARGET_ARM64
+                else if (assignedInterval->getNextRefPosition()->needsConsecutive)
+                {
+                    // if next refposition is part of consecutive registers and there is already a register
+                    // assigned to it then do not reassign for currentRefPosition, because with that, other
+                    // registers for the next consecutive register assignment would have to be copied to
+                    // different consecutive registers since this register is busy from this point onwards.
+                    continue;
```
Comment (BruceForstall) [inline src/coreclr/jit/lsra.cpp:11973]: Would it be possible to run out of registers if there are a number of instrinsics defining consecutive registers and later using them, and in between we don't allow spilling them?

**Review by BruceForstall — APPROVED** [review]

File: src/coreclr/jit/lsra.cpp (hunk 1 above)
Comment (kunalspathak) [reply src/coreclr/jit/lsra.cpp:11973]: Yes, that was exactly on my mind and hence wanted to check if we find a case like that in CI. I think I will revert this portion.

Comment (kunalspathak) [reply src/coreclr/jit/lsra.cpp:11973]: On a second thought, I could actually have it record the spill cost as costlier, so those registers will be preferred if no other option is available. With that we do eliminate the extra copies and the code looks much better:

main vs. latest changes

image


skip spilling consecutive vs. latest changes

image

**Pull request merged**